Index: scripts/check_jni_methods.sh.in
===================================================================
RCS file: /sources/classpath/classpath/scripts/check_jni_methods.sh.in,v
retrieving revision 1.5
diff -u -r1.5 check_jni_methods.sh.in
--- scripts/check_jni_methods.sh.in	11 Jun 2008 21:37:49 -0000	1.5
+++ scripts/check_jni_methods.sh.in	23 Jun 2008 20:03:20 -0000
@@ -6,12 +6,16 @@
 TMPFILE=/tmp/check-jni-methods.$$.1
 TMPFILE2=/tmp/check-jni-methods.$$.2
 TMPFILE3=/tmp/check-jni-methods.$$.3
+TMPFILE4=/tmp/check-jni-methods.$$.4
 
 # Find all methods defined in the header files generated
 # from the java source files.
-grep -h '^JNIEXPORT .* Java_' @abs_top_builddir@/include/*.h @abs_top_srcdir@/include/*.h | \
-        LC_ALL=C sed -e 's,.*JNICALL \(Java_[a-z_A-Z0-9]*\).*$,\1,' | \
-	sort -u > $TMPFILE
+grep -h '^JNIEXPORT .* Java_' @abs_top_srcdir@/include/*.h | \
+        LC_ALL=C sed -e 's,.*JNICALL \(Java_[a-z_A-Z0-9]*\).*$,\1,' > $TMPFILE
+grep -h '^JNIEXPORT .* Java_' @abs_top_builddir@/include/*.h | \
+        LC_ALL=C sed -e 's,.*JNICALL \(Java_[a-z_A-Z0-9]*\).*$,\1,' >> $TMPFILE
+sort -u $TMPFILE >> $TMPFILE4
+mv $TMPFILE4 $TMPFILE
 
 # Find all methods in the JNI C source files.
 find @abs_top_srcdir@/native/jni -name \*.c | \
