--- java/util/PriorityQueue.java	2006-12-22 00:02:50.000000000 +1100
+++ java/util/PriorityQueue.java	2010-05-02 22:36:24.000000000 +1000
@@ -108,6 +108,7 @@
 
   public PriorityQueue(int cap, Comparator<? super E> comp)
   {
+    if (cap <= 0) throw new IllegalArgumentException();
     this.used = 0;
     this.storage = (E[]) new Object[cap];
     this.comparator = comp;
@@ -170,6 +173,9 @@
       public void remove()
       {
 	PriorityQueue.this.remove(index);
+	// removal promotes another element to the slot just vacated,
+	// so we need to rewind to pick up the newly promoted element.
+	--index;
       }
     };
   }
