--- native/jni/native-lib/cpio.c	2007-02-09 21:51:07.000000000 +0200
+++ native/jni/native-lib/cpio.c	2007-08-21 23:52:15.352763000 +0300
@@ -526,7 +526,12 @@
   struct dirent *dBuf;
 
 #ifdef HAVE_READDIR_R
+#if defined (__SVR4) && defined (__sun)
+  /*The solaris version of readdir_r is somewhat different*/
+  dBuf = readdir_r((DIR *)handle, &dent);
+#else
   readdir_r ((DIR *) handle, &dent, &dBuf);
+#endif /*__sun__*/
 #else
   dBuf = readdir((DIR *)handle);
 #endif /* HAVE_READDIR_R */
