# This is GNU Classpath commit c6649596a7cf598c5d382f973ff5d7cecb255790 .
diff --git a/native/jni/gtk-peer/gnu_java_awt_peer_gtk_GdkFontPeer.c b/native/jni/gtk-peer/gnu_java_awt_peer_gtk_GdkFontPeer.c
index a5e59f7..771b23e 100644
--- a/native/jni/gtk-peer/gnu_java_awt_peer_gtk_GdkFontPeer.c
+++ b/native/jni/gtk-peer/gnu_java_awt_peer_gtk_GdkFontPeer.c
@@ -59,14 +59,14 @@ enum java_awt_font_baseline {
   java_awt_font_HANGING_BASELINE = 2
 };
 
-static PangoFT2FontMap *ft2_map = NULL;
+static PangoFontMap *font_map = NULL;
 
 JNIEXPORT void JNICALL
 Java_gnu_java_awt_peer_gtk_GdkFontPeer_initStaticState 
   (JNIEnv *env, jclass clazz __attribute__((unused)))
 {
   gtkpeer_init_font_IDs(env);
-  ft2_map = PANGO_FT2_FONT_MAP(pango_ft2_font_map_new());
+  font_map = pango_ft2_font_map_new();
 }
 
 JNIEXPORT void JNICALL
@@ -287,7 +287,7 @@ Java_gnu_java_awt_peer_gtk_GdkFontPeer_setFont
   pango_font_description_set_size (pfont->desc, size * PANGO_SCALE);
   
   /* Create new context */
-  pfont->ctx = pango_ft2_font_map_create_context (ft2_map);
+  pfont->ctx = pango_font_map_create_context (font_map);
   g_assert (pfont->ctx != NULL);
   
   pango_context_set_font_description (pfont->ctx, pfont->desc);
