--- native/jni/gtk-peer/gnu_java_awt_peer_gtk_GtkToolkit.c	2007-04-04 12:30:32.000000000 +0100
+++ native/jni/gtk-peer/gnu_java_awt_peer_gtk_GtkToolkit.c	2007-09-27 13:48:41.000000000 +0100
@@ -182,10 +182,6 @@
 
   gtk_widget_set_default_colormap (gdk_rgb_get_colormap ());
 
-  /* Make sure queued calls don't get sent to GTK/GDK while 
-     we're shutting down. */
-  atexit (gdk_threads_enter);
-
   if ((homedir = getenv ("HOME")))
     {
       rcpath = (char *) g_malloc (strlen (homedir) + strlen (RC_FILE) + 2);
