--- scripts/check_jni_methods.sh.in	2008-06-01 18:09:53.000000000 -0400
+++ scripts/check_jni_methods.sh	2008-06-06 12:07:28.000000000 -0400
@@ -9,20 +9,20 @@
 
 # Find all methods defined in the header files generated
 # from the java source files.
-grep -h '^JNIEXPORT .* Java_' @abs_top_builddir@/include/*.h @abs_top_srcdir@/include/*.h | \
+grep -h '^JNIEXPORT .* Java_' /home/dgrove/rvm-trunk/components/classpath/97.2p1/classpath/include/*.h /home/dgrove/rvm-trunk/components/classpath/97.2p1/classpath/include/*.h | \
         LC_ALL=C sed -e 's,.*JNICALL \(Java_[a-z_A-Z0-9]*\).*$,\1,' | \
 	sort -u > $TMPFILE
 
 # Find all methods in the JNI C source files.
-find @abs_top_srcdir@/native/jni -name \*.c | \
+find /home/dgrove/rvm-trunk/components/classpath/97.2p1/classpath/native/jni -name \*.c | \
 	xargs grep -h '^Java_' | \
         LC_ALL=C sed -e 's,^\(Java_[a-z_A-Z0-9]*\).*$,\1,' > $TMPFILE2
 # Or in the the C++ files. (Note that cpp doesn't follow gnu conventions atm)
 # So we try to match both GNU style and some other style.
-find @abs_top_srcdir@/native/jni -name \*.cpp | \
+find /home/dgrove/rvm-trunk/components/classpath/97.2p1/classpath/native/jni -name \*.cpp | \
 	xargs grep -h '^Java_' | \
         LC_ALL=C sed -e 's,^\(Java_[a-z_A-Z0-9]*\).*$,\1,' >> $TMPFILE2
-find @abs_top_srcdir@/native/jni -name \*.cpp | \
+find /home/dgrove/rvm-trunk/components/classpath/97.2p1/classpath/native/jni -name \*.cpp | \
 	xargs egrep -h '^(JNIEXPORT .* JNICALL )?Java_' | \
 	cut -f4 -d\  | \
         LC_ALL=C sed -e 's,^\JNIEXPORT .* JNICALL \(Java_[a-z_A-Z0-9]*\).*$,\1,' >> $TMPFILE2
